spec.bs

```diff
@@ -0,0 +1,520 @@
+
+  Note: In Chrome’s experimentation phase, the [=host=] of a {{Document}}'s {{URL}} is used as the [=determine-topics-calculation-input-data-header/topics calculation input data=], and the model is trained with human curated hostnames and topics.
+
+  <h2 id="collect-page-topics-calculation-input-data-header">Collect page topics calculation input data</h2>
+  <div algorithm>
+    To <dfn>collect page topics calculation input data</dfn>, given a {{Document}} |document|:
+
+    1. If |document|'s [=node navigable=] is a [=prerendering navigable=], then append the following steps to |document|'s [=post-prerendering activation steps list=] and return. Else, run the following steps [=in parallel=]:
+      1. Let |documentId| be |document|'s [=document-id-header/document ID=].
+      1. If user agent's [=user agent/topics history storage=] contains a [=topics history entry=] whose [=topics history entry/document id=] is |documentId|, return.
+      1. Let |topicsHistoryEntry| be a [=topics history entry=].
+      1. Set |topicsHistoryEntry|'s [=topics history entry/document id=] to |documentId|.
+      1. Set |topicsHistoryEntry|'s [=topics history entry/topics calculation input data=] to the [=determine-topics-calculation-input-data-header/topics calculation input data=] for |document|.
+      1. Set |topicsHistoryEntry|'s [=topics history entry/time=] to |document|'s [=relevant settings object=]'s [=current relative timestamp=].
+      1. [=list/Append=] |topicsHistoryEntry| to user agent's [=user agent/topics history storage=].
+  </div>
+
+  <h2 id="collect-topics-caller-origin-header">Collect topics caller origin</h2>
+  <div algorithm>
+    To <dfn>collect topics caller origin</dfn>, given a {{Document}} |document| and an [=origin=] |callerOrigin|:
+
+    1. Run the following steps [=in parallel=]:
+      1. Let |documentId| be |document|'s [=document-id-header/document ID=].
+      1. If user agent's [=user agent/topics history storage=] does not contain a [=topics history entry=] whose [=topics history entry/document id=] is |documentId|, return.
+      1. Let |topicsHistoryEntry| be the [=topics history entry=] in user agent's [=user agent/topics history storage=] whose [=topics history entry/document id=] is |documentId|.
+      1. [=set/Append=] |callerOrigin| to |topicsHistoryEntry|'s [=topics caller origins=].
+  </div>
+
+  <h2 id="periodically-calculate-user-topics-header">Periodically calculate user topics</h2>
+  <div algorithm>
+    To <dfn>calculate user topics</dfn>, perform the following steps every 7 days:
+    1. Let |now| be the [=monotonic clock=]'s [=monotonic clock/unsafe current time=].
+    1. If either user agent's [=user agent/model=] or [=user agent/taxonomy=] isn't available:
+        1. Let |epoch| be an [=epoch=] struct with default initial field values.
+        1. Set |epoch|'s [=epoch/time=] to |now|.
```

Comment:
Type error: the time is a "[DOMHighResTimeStamp](https://w3c.github.io/hr-time/#dom-domhighrestimestamp)", but |now| is an unsafe moment. If you change the epoch's time, you'll have to push that through to several places that compare it against a duration or that concatenate it with strings. If you change |now|'s type, you'll need to note what duration the epoch's time is measuring, and again make sure that's consistent in all the places you're using it.
You can use '[coarsen](https://w3c.github.io/hr-time/#dfn-coarsen-time)' directly without having an environment settings object, if you don't care about the sub-millisecond resolution that it removes.

Reply:
"If you change |now|'s type, you'll need to note what duration the epoch's time is measuring". Could you clarify? I changed |now| to the "coarsen" of wall clock’s current time, and then it seems all the time references are consistent & comparable. I didn't see any notion of "time origin" in those process (or are they implicitly from "Unix epoch"?)

Reply:
I figured I just shouldn’t use any of "the document’s relevant settings object's current relative timestamp" as those times aren’t performance related and they won't be compared within the page. Instead, I'm using a single consistent way to get the current time, and keep using Unix epoch as the time origin. PTAL again.